torch/lib/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1413,6 +1413,15 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupNCCL::barrier(
     // ensure that each process is on a different GPU
     auto numGPUs = at::cuda::getNumGPUs();
     int16_t deviceIdx = static_cast<int16_t>(rank_ % numGPUs);
+    auto bestGuessWarning = c10::str(
+      "Rank ",
+      this->getRank(),
+      " using best-guess GPU ",
+      deviceIdx,
+      " to perform barrier as devices used by this process are currently unknown. ",
+      "Specify device_ids in barrier() to force use of a particular device."
```

Comment:
Better to say "To avoid any potential hang due to a wrong guess, please specify ...". This can improve the debuggability by mentioning the potential outcome if the warning here is ignored.